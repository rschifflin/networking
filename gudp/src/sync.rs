use std::sync::{Mutex, MutexGuard, PoisonError, Condvar};
use std::ops::{Deref, DerefMut};
use std::fmt;

/// Mutex with associated condvar.
/// The condvar can only be used with the associated mutex guard,
/// and only when the associated lock is held.
///
/// IE instead of (Mutex<resource>, Condvar), use CondMutex<resource>
/// This is a convenience wrapper and thus does not export
/// the entire api of Mutex. If the whole api is needed,
/// consider calling `parts()`
///
/// CondMutex also has an optional (preferably zero-sized) tag type marker,
/// to allow type-level naming of the resource guarded by the mutex.
/// For example, given a mutex-wrapped reader vec and a mutex-wrapped writer vec,
/// a reader fn could take CondMutex<Vec, READER_TAG_TYPE> to prevent
/// accidentally passing the writer vec which is otherwise identical to the typesystem.
pub type LockError<'a, T> = PoisonError<MutexGuard<'a, T>>;

#[derive(Debug, Default)]
pub struct CondMutex<T, Tag: Copy = ()> {
  _tag: Tag,
  mx: Mutex<T>,
  cv: Condvar
}

impl<T> CondMutex<T> {
  pub fn new(t: T) -> CondMutex<T> {
    CondMutex {
      _tag: (),
      mx: Mutex::new(t),
      cv: Condvar::new()
    }
  }

  pub fn parts(&self) -> (&Mutex<T>, &Condvar) {
    (&self.mx, &self.cv)
  }

  pub fn parts_mut(&mut self) -> (&mut Mutex<T>, &Condvar) {
    (&mut self.mx, &self.cv)
  }

  pub fn into_parts(self) -> (Mutex<T>, Condvar) {
    (self.mx, self.cv)
  }

  pub fn lock(&self) -> Result<CondMutexGuard<T>, LockError<T>> {
    self.mx.lock()
      .map(|guard| CondMutexGuard { _tag: self._tag, guard, cv: &self.cv })
  }
}

#[derive(Debug)]
pub struct CondMutexGuard<'a, T: ?Sized + 'a, Tag: Copy = ()> {
  _tag: Tag,
  guard: MutexGuard<'a, T>,
  cv: &'a Condvar
}

impl <'a, T> CondMutexGuard<'a, T> {
  pub fn parts(&self) -> (&MutexGuard<'a, T>, &Condvar) {
    (&self.guard, self.cv)
  }

  // NOTE: Condvar api is entirely immutable, so we only provide immutable refs
  pub fn parts_mut(&mut self) -> (&mut MutexGuard<'a, T>, &Condvar) {
    (&mut self.guard, self.cv)
  }

  pub fn into_parts(self) -> (MutexGuard<'a, T>, &'a Condvar) {
    (self.guard, self.cv)
  }

  pub fn wait(self) -> Result<CondMutexGuard<'a, T>, LockError<'a, T>> {
    let _tag = self._tag;
    let cv = self.cv;
    let guard = self.guard;
    let res = cv.wait(guard);

    res.map(|guard| CondMutexGuard { _tag, guard, cv })
  }

  pub fn notify_one(&self) {
    self.cv.notify_one()
  }

  pub fn notify_all(&self) {
    self.cv.notify_all()
  }
}

impl<T: ?Sized> Deref for CondMutexGuard<'_, T> {
  type Target = T;

  fn deref(&self) -> &T {
    self.guard.deref()
  }
}

impl<T: ?Sized> DerefMut for CondMutexGuard<'_, T> {
  fn deref_mut(&mut self) -> &mut T {
    self.guard.deref_mut()
  }
}

impl<T: ?Sized + fmt::Display> fmt::Display for CondMutexGuard<'_, T> {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    (**self).fmt(f)
  }
}
