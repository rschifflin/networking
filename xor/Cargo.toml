[package]
name = "xor"
version = "0.1.0"
authors = ["Mark Schifflin <randy.schifflin@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dev-dependencies]
trybuild = "1.0"
