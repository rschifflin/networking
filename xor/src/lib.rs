mod xor;
mod r#mut;
mod owned;

pub use crate::xor::Xor;
pub use crate::r#mut::Xor as XorMut;
pub use crate::owned::Xor as XorOwned;

#[test]
fn test_compiler_disallows_accessing_both() {
  let t = trybuild::TestCases::new();
  t.compile_fail("tests/disallow/*.rs");
}

/*

#[test]
fn test_disallow_borrowing_both_mut() {
  let lhs = vec![3usize];
  let rhs = vec![5usize];
  let both = &mut (lhs, rhs);
  let mut xor = XorMut::new(both);

  let hold_left = xor.lhs();
  let hold_right = xor.rhs();
  assert_eq!(hold_left.pop(), Some(3usize));
  assert_eq!(hold_right.pop(), Some(5usize));
}

#[test]
fn test_disallow_borrowing_both_arc_mutex() {
  use std::sync::Arc;
  let lhs = vec![8usize];
  let rhs = vec![1usize];
  let both = Arc::new((lhs, rhs));
  let mut xor = Xor::new(both);

  let hold_left = xor.lhs();
  let hold_right = xor.rhs();
  assert_eq!(hold_left.pop(), Some(8usize));
  assert_eq!(hold_right.pop(), Some(1usize));
}
*/
