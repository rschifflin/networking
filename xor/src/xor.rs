use std::ops::Deref;

/// A reference to a tuple holding T and U, but preventing borrowing both at once.
pub struct Xor<T,U, Both: Deref<Target=(T,U)>> {
  pair: Both
}

impl<T,U,Both: Deref<Target=(T,U)>> Xor<T,U,Both> {
  pub fn new(pair: Both) -> Xor<T,U,Both> {
    Xor { pair }
  }

  pub fn lhs(&mut self) -> &T {
    &self.pair.0
  }

  pub fn rhs(&mut self) -> &U {
    &self.pair.1
  }
}

#[test]
fn test_ref() {
  let lhs = vec![1usize];
  let rhs = vec![2usize];
  let both = &(lhs, rhs);
  let mut xor = Xor::new(both);
  assert_eq!(xor.lhs().as_slice(), &[1usize]);
  assert_eq!(xor.rhs().as_slice(), &[2usize]);
}

#[test]
fn test_manual_drop_reallows() {
  use std::sync::Mutex;

  let lhs = Mutex::new(false);
  let rhs = Mutex::new(false);
  let both = &(lhs, rhs);
  let mut xor = Xor::new(both);

  let mut lhs = xor.lhs().lock().unwrap();
  *lhs = true;
  drop(lhs);
  let mut rhs = xor.rhs().lock().unwrap();
  *rhs = true;
}

#[test]
fn test_with_arc() {
  use std::sync::Arc;

  let lhs = vec![8usize];
  let rhs = vec![1usize];
  let both = Arc::new((lhs, rhs));
  let mut xor = Xor::new(both);
  assert_eq!(xor.lhs().first(), Some(&8usize));
  assert_eq!(xor.rhs().first(), Some(&1usize));
}
