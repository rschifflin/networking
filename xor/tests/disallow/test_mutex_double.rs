use std::sync::Mutex;
use xor::Xor;

fn main() {
  let lhs = Mutex::new(false);
  let rhs = Mutex::new(false);
  let both = &(lhs, rhs);
  let mut xor = Xor::new(both);

  let mut lhs = xor.lhs().lock().unwrap();
  *lhs = true;

  let mut rhs = xor.rhs().lock().unwrap();
  *rhs = true;
}
