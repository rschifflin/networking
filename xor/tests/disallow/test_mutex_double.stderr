error[E0499]: cannot borrow `xor` as mutable more than once at a time
  --> $DIR/test_mutex_double.rs:13:17
   |
10 |   let mut lhs = xor.lhs().lock().unwrap();
   |                 --- first mutable borrow occurs here
...
13 |   let mut rhs = xor.rhs().lock().unwrap();
   |                 ^^^ second mutable borrow occurs here
14 |   *rhs = true;
15 | }
   | - first borrow might be used here, when `lhs` is dropped and runs the `Drop` code for type `MutexGuard`
